### ISSUE TITLE
Add P2P communication with MySearch, Seeks and YaCy nodes

### ISSUE THREAD

**Issue by GreenLunar** [searx/searx#393]
Add peer-to-peer communication with Seeks nodes, if is not already possible.

It would be great.  When some decentralized search engine platforms would add mutuality features with other decentralized search engine platforms, we might be able to form some sort of standard to manage bibliographical references.

**_Bibliographical database search library**_

_EndNote is a pretty heavy-duty bibliography manager for Mac and Windows, with a heavy-duty cost attached to it. But really it's not doing much more than providing a handy interface to online bibliographical databases. It would be nice to see an extensible libbib. It even rhymes._  From [Project ideas - Movementarian.org](http://movementarian.org/code/projectideas/)

**Comment by jpope777** [comment]
As a Seeks->Searx convert, are we sure there are enough Seeks nodes still running these days for this to be applicable? 

Even before my node finally died, I noticed that 99% of the nodes that I was peered with were no longer in existence...

**Comment by GreenLunar** [comment]
I think there are enough, there are, at least, five public [active Seeks nodes](https://github.com/beniz/seeks/wiki/List-of-Web-Seeks-nodes#Trusted_Nodes) available.

**Comment by Cqoicebordel** [comment]
3 . The others are error 404 or 403.

**Comment by GreenLunar** [comment]
https://seeks.mxchange.org/ should be https://seeks.mxchange.org/search.php

See https://github.com/beniz/seeks/issues/16#issuecomment-121771057

**Comment by dalf** [comment]
See Pull Request #6 : that's suppose to record the links that are clicked by the user.
For now searx is stateless, with the seeks protocol support this won't be the case any more.

Another thing : the seeks protocol has to be review to make sure it is compatible with searx spirit. 

One way to make everyone happy : implement everything as a plugin (but this required to change / improve the plugin system).

**Comment by GreenLunar** [comment]
> For now searx is stateless, with the seeks protocol support this won't be the case any more.

I disagree, I think you can have it both ways.  We can, as you pointed out, have it implemented as a plugin or as a switch (true/false configuration option).

**Comment by Cqoicebordel** [comment]
The issue is not that. Stateless means that searx doesn't remember _anything_ : you can kill searx and relaunch it, and you will find the same results. But, for connecting it with a seeks node, it needs a long term memory. That will change _a lot_ of the way searx is working right now.

**Comment by dalf** [comment]
Idea : have two servers, one stateless, another one statefull.

How to do that ? 

Add autonomous server / part : 
- which does the p2p communication
- which can behave as a search engine
- which can record user choice for a request (not the web part, only an API)

Improve the plugin system to be able to add some public URL : this would allow to record the URL the user has clicked. Internally, the choice would be recorded using the API on the p2p server.

The protocol between the web side and the p2p server has to be choosen (HTTP or protocol buffer, or ?)

Side effect : technically, we could have many web instance for one p2p server. Not sure this is the purpose.

Start of the out of topic thoughts : the more time I spend time thinking about searx, the more I believe that we should have one process and different threads instead of multiple process which are independant.

Examples of use : 
- sharing the statistics : it can be done with Redis / an ORM, but easier if it is in memory. The project is not storing gigabytes of statistics, only few numbers.
- sharing HTTP(s) connection to other server.
- In my opinion, we should have something that dispatch all the external requests taking them as one flow. For example : the PR that does asynchronous parsing of result create multiple thread, and this happen for each process, in fact this doesn't scale at all. Either we use a framework that doesn't need one thread per HTTP requests, either we have one unique pool of threads for the whole searx server.

The oposite idea is to have multiple independant searx communicating each other, even on one physical server. In this case the p2p / could is native. But I think it is overcomplicated for simple needs. 

I don't know enough python, but we should give a try to Twisted ?

End of out of topic

**Comment by GreenLunar** [comment]
Foren-Übersicht ‹ YaCy Development ‹ Jobs

get YaCy to connect to the outside world
Beitragvon Orbiter » Mi Jul 10, 2013 9:09 am

http://forum.yacy-websuche.de/viewtopic.php?f=21&t=4702
http://www.freelancer.com/projects/System-Admin-Troubleshooting/Help-get-YaCy-connect-the.html

Help me get YaCy to connect to the outside world not reaching any peers
_This project received 7 bids from talented freelancers with an average bid price of $22 USD / hour._

@yacy, are you interested in YaCy to be connected with searx and Seeks nodes?

**Comment by HLFH** [comment]
There is also now the MySearch project on the open forge CodingTeam: https://codingteam.net/project/mysearch 

So I suggest we also add P2P communication with MySearch.

**Comment by GreenLunar** [comment]
Good idea.

Relevant link https://codingteam.net/project/mysearch/bugs/show/3053

**Comment by GreenLunar** [comment]
Relevant link of a counter issue https://codingteam.net/project/mysearch/bugs/show/3053

**Comment by GreenLunar** [comment]
It would be great.  When some decentralized search engine platforms would add mutuality features with other decentralized search engine platforms, we might be able to form some sort of standard to manage bibliographical references.

***Bibliographical database search library***

*EndNote is a pretty heavy-duty bibliography manager for Mac and Windows, with a heavy-duty cost attached to it. But really it's not doing much more than providing a handy interface to online bibliographical databases. It would be nice to see an extensible libbib. It even rhymes.*  From [Project ideas - Movementarian.org](http://movementarian.org/code/projectideas/)

**Comment by seniorm0ment** [comment]
Big +1 to this
